Fixes a bunch of revenant stuff, again.

ChangelingRain opened the pull request:
Splits Defile into two abilities, Defile and Malfunction.

> Harvest now can be activated by clicking the target while in range.
> Defile affects a 7x7 area, damaging windows, tiring humans, and corrupting tiles.
> Malfunction affects a 9x9 area, emping it, emagging bots, and possibly(10%) emagging machines.
> Defile and Overload Lights now briefly stun you so you can't just trigger and run instantly.
> Fixed a bunch of Harvest bugs.

ChangelingRain committed:
Merge pull request #3 from Anonus/fixrevenants

FUCK GIT

ChangelingRain committed:
Splits defile into two abilities, Defile and Malfunction. Harvest now can be activated by clicking the target while in range.
Defile affects a 7x7 area, damaging windows, tiring humans, and corrupting tiles.
Malfunction affects a 9x9 area, emagging bots, disabling cyborgs, and emping or emagging machines.
Defile and Overload Lights now briefly stun you so you can't just trigger and run instantly.
Fixed a bunch of Harvest bugs.

ChangelingRain committed:
Changelog, changelogs are fun!

ChangelingRain committed:
there was no bug here

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -78,6 +81,77 @@
+/mob/living/simple_animal/revenant/proc/Harvest(mob/living/carbon/human/tar, mob/living/simple_animal/revenant/user = usr)
+	if(!user.castcheck(0))
+		return
+	var/mob/living/carbon/human/target = tar
```
why. ```tar``` is already human, just do ```if(istype(tar))``` and then you can use ```tar``` from then on.
if you don't like it being called ```tar``` (i don't) then change it's name in the arguments for ```Harvest()```

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -78,6 +81,77 @@
+		Harvest(A, src)
+
+
+/mob/living/simple_animal/revenant/proc/Harvest(mob/living/carbon/human/tar, mob/living/simple_animal/revenant/user = usr)
```
Remove the user argument of this.
it's a revenant proc, you are NOT adding a revenant argument to it because that's stupid.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -1,102 +1,29 @@
-			user << "<span class='info'>[target]'s soul has been considerably weakened and will yield no more essence for the time being.</span>"
-			drained_mobs.Add(target)
-			draining = 0
+			user.Harvest(target, user)
```
```user.Harvest(target)```

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant_abilities.dm:
```diff
@@ -185,22 +117,71 @@
-			for(var/obj/machinery/light/light in T.contents)
-				light.flicker() //spooky
-	user.reveal(100)
+			for(var/mob/living/carbon/human/human in T.contents)
```
it might be easier to just spawn the emp effect, since I believe that already covers all the logic of ensuring everything nearby is emp'd

Anonus replied:
Well, right now it doesn't mess with locker or crate locks.
If I want it to randomly unlock/relock lockers and crates(and damage the AI directly) I'll change it to an actual EMP, but there are advantages of making it like this, such as being able to exclude stuff from the effect(such as, perhaps, APCs, so that it can't murder the AI while the AI can do nothing.)

RemieRichards replied:
yes but Malfunction is 50E and 15E per use, by that point the revenant has earned the ability to kill the AI, in my eyes, and really, besides being a dick, it has no need to (as the AI can't see the revenant unless people can, and people are the only thing it can drain)

plus this is a very snowflakey almost copy-pasta implementation of EMP effects at this point.

Anonus replied:
Yeah sure fuck it you've got points I mostly agree with

RemieRichards commented:
If what you thought was ugly last night was all the ```if(target)``` checks, yes, that is ugly, but you should see ```do_after()``` (In fact, perhaps swap those ```sleep()``` with ```do_after()``` as you're repeating a lot of the same logic)

Robustin commented:
Revenants are already super annoying and hard to kill, if malfunction is going to emag doors to I might as well call the shuttle when one shows up.

ChangelingRain committed:
YOU SPEAK, I OBEY. Fixes the stuff, as ordered, at least hopefully in the right way.
Tweaks Malfunction a bit to reduce the chance of random emagging; it'll mostly be emp_act now.

Anonus commented:
It'll have a low chance to emag(or at least, attempt to), set at 10% right now. The rest of the time, it'll EMP stuff.
Also yeah they're a bit flighty, so they're briefly, or not so briefly, stunned when using dangerous abilities.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -78,6 +81,72 @@
 		return
 	A.attack_ghost(src)
+	if(ishuman(A) && in_range(src, A))
+		src.Harvest(A)
```
if ```src``` is the revenant (which is most of these cases I believe it is) then ```src``` is implied, please remove all ```src``` where possible.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -78,6 +81,72 @@
+				if(5 to INFINITY)
+					src << "<span class='info'>Such a feast! [target] will yield much essence to you.</span>"
+			if(do_after(src, 30, 9, 0, target)) //how about now
+				if(!in_range(src, target))
```
pretty sure ```do_after()``` already checks that the user (src) and the target (target) are in range.

ChangelingRain committed:
Malfunction now uses empulse instead of copypasta code.
Reduces how much power gets drained from the APC when a machine is EMPed; right now, four machines could fully discharge an APC.
Obeys the Remie, no unneeded src.
Slightly upped the stun on defile and overload light(go slower, revenants!)

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -196,21 +266,33 @@
+
+/mob/living/simple_animal/revenant/proc/stun(time)
+	var/mob/living/simple_animal/revenant/R = usr
+	if(!istype(usr) || !usr)
```
```!usr``` is unnecessary, if ```usr``` is ```null``` it will also fail the ```istype()```, meaning the ```if(!usr)``` part is never called.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -196,21 +266,33 @@
-/mob/living/simple_animal/revenant/proc/reveal(time, stun)
+/mob/living/simple_animal/revenant/proc/reveal(time)
 	var/mob/living/simple_animal/revenant/R = usr
 	if(!istype(usr) || !usr)
```
see below

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -196,21 +266,33 @@
+		R << "<span class='notice'>You are once more concealed.</span>"
+
+/mob/living/simple_animal/revenant/proc/stun(time)
+	var/mob/living/simple_animal/revenant/R = usr
```
typecasting usr? no, use src, implied.

RemieRichards commented on code/modules/mob/living/simple_animal/revenant/revenant.dm:
```diff
@@ -196,21 +266,33 @@
 
-/mob/living/simple_animal/revenant/proc/reveal(time, stun)
+/mob/living/simple_animal/revenant/proc/reveal(time)
 	var/mob/living/simple_animal/revenant/R = usr
```
see below

ChangelingRain committed:
Fixes xhuis's shit. Also fixes a bug where you would use essence if you tried to use an ability while inhibited.

ChangelingRain committed:
Wait shit that shouldn't be indented~

ChangelingRain committed:
Harvest ability is dead and gone, just click them, nerd.
